onnxruntime/core/providers/cpu/tensor/upsample.cc

```diff
@@ -66,30 +39,122 @@ Status UpsampleNearest(const T* input,
     return Status(ONNXRUNTIME, FAIL, "Upsample: input/output value is nullptr");
   if (input_shape.NumDimensions() != output_shape.NumDimensions())
     return Status(ONNXRUNTIME, FAIL, "Upsample: input/output value's dimension mismatch");
-  auto n_dim = input_shape.NumDimensions();
-  if (scales.size() == 4 && scales[0] == 1 && scales[1] == 1 && scales[2] == 2 && scales[3] == 2) {
-    UpsampleNearest2x<T>(input_shape[0], input_shape[1], input_shape[2], input_shape[3], input, output);
-  } else {
-    for (size_t i = 0, size = output_shape.Size(); i < size; i++) {
-      size_t old_idx = 0;
-      size_t cur_idx = i;
+  if (input_shape.NumDimensions() == 0) {
+    return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
+                  "Upsample: input shape needs to be at least a single dimension.");
+  }
 
-      int64_t base = 1;
-      for (auto j = static_cast<int64_t>(n_dim - 1); j >= 0; j--) {
-        auto tmp = cur_idx % output_shape[j];
+  int64_t n_dim = static_cast<int64_t>(input_shape.NumDimensions());
+  std::vector<int64_t> output_dim_counter(n_dim);
+  output_dim_counter[n_dim - 1] = -1;  // initialize dimension counter
+
+  std::vector<int64_t> input_dim_counters(n_dim);
+  std::vector<int64_t> input_dim_factor(n_dim);
+  input_dim_factor[n_dim - 1] = 1;  // initialize dimension factor
+  for (int64_t dim_idx = n_dim - 2; dim_idx >= 0; dim_idx--) {
+    input_dim_factor[dim_idx] = input_dim_factor[dim_idx + 1] * input_shape[dim_idx + 1];
+  }
+
+  std::vector<int64_t> output_dim_factor(n_dim);
+  output_dim_factor[n_dim - 1] = 1;  // initialize dimension factor
+  for (int64_t dim_idx = n_dim - 2; dim_idx >= 0; dim_idx--) {
+    output_dim_factor[dim_idx] = output_dim_factor[dim_idx + 1] * output_shape[dim_idx + 1];
+  }
+
+  int64_t output_idx = 0;
+  int64_t input_idx = 0;
+
+#define OneDemensionProcessor(dim_inx)                                                                                                                  \
```

Comment:
Would you measure how much the second opt improve in comparison to your first opt? Thanks.

Reply:
It has about 3x gain for the input case in the description section. opt 1 takes ~220 ms, opt 2 takes ~75ms.